## This will be a brief example of a CNN using Keras, Tensorflow etc

In [2]:
# Import the MNIST data from Keras
from keras.utils import np_utils
from keras.models import Sequential
from keras.layers.core import Dense, Dropout, Activation, Flatten
from keras.layers.convolutional import Convolution2D, MaxPooling2D
from sklearn.model_selection import train_test_split
import pandas as pd
import numpy as np

Using TensorFlow backend.
C:\ProgramData\Anaconda3\lib\site-packages\tensorboard\compat\tensorflow_stub\dtypes.py:541: FutureWarning: Passing (type, 1) or '1type' as a synonym of type is deprecated; in a future version of numpy, it will be understood as (type, (1,)) / '(1,)type'.
  _np_qint8 = np.dtype([("qint8", np.int8, 1)])
C:\ProgramData\Anaconda3\lib\site-packages\tensorboard\compat\tensorflow_stub\dtypes.py:542: FutureWarning: Passing (type, 1) or '1type' as a synonym of type is deprecated; in a future version of numpy, it will be understood as (type, (1,)) / '(1,)type'.
  _np_quint8 = np.dtype([("quint8", np.uint8, 1)])
C:\ProgramData\Anaconda3\lib\site-packages\tensorboard\compat\tensorflow_stub\dtypes.py:543: FutureWarning: Passing (type, 1) or '1type' as a synonym of type is deprecated; in a future version of numpy, it will be understood as (type, (1,)) / '(1,)type'.
  _np_qint16 = np.dtype([("qint16", np.int16, 1)])
C:\ProgramData\Anaconda3\lib\site-packages\tensorboard\comp

In [7]:
np._module

AttributeError: module 'numpy' has no attribute '_module'

In [2]:
# Define the test and training datasets
from keras.datasets import mnist
(X_train, y_train), (X_test, y_test) = mnist.load_data()

# All the data are loaded now. They need to be reshaped into 1D vectors for the first layer
X_train.shape

(60000, 28, 28)

In [3]:
# Preprocessing

# Reshape the data
X_train = X_train.reshape(60000,28,28,1)
X_test = X_test.reshape(10000,28,28,1)

# One hot encode the labels
y_train = np_utils.to_categorical(y_train)
y_test = np_utils.to_categorical(y_test)

In [4]:
y_train[0]  # If this doesn't look right, check that you haven't accidentally run the prior block multiple times

array([0., 0., 0., 0., 0., 1., 0., 0., 0., 0.], dtype=float32)

We now define a model comprised of two dimensional convolutional layers, each with a 'pooling' layer following it.

First layer uses 64 nodes, second layer uses 32, and the convolution kernel (filter) is 3 pixels by 3 pixels.

The 'flatten' layer flattens the inputs from a matrix into a vector

The dense layer takes the flattened vector and 'generates probabilities for 10 target labels using the softmax activation function'

In [7]:
model = Sequential()

model.add(Conv2D(64, kernel_size=3, activation='relu', input_shape=(28,28,1)))

model.add(Conv2D(32, kernel_size=3, activation='relu'))

model.add(Flatten())

model.add(Dense(10, activation='softmax'))  # This is the back end? The probabilities are the ones that are thresholded
# and converted into predictions for what the image is?

model._layers

Now need to compile the model. This requires three other pieces of info: The loss function to be minimised, the optimiser that'll do the minimising, and the metric by which it minimises

In [12]:
model.compile(optimizer='adam', loss='categorical_crossentropy', metrics=['accuracy'])

In [13]:
# Train the model. Be aware that this'll take a while (~15 minutes on my laptop for 3 epochs over 60,000 training images)
model.fit(X_train, y_train, validation_data=(X_test, y_test), epochs=3)

Instructions for updating:
Use tf.where in 2.0, which has the same broadcast rule as np.where

Train on 60000 samples, validate on 10000 samples
Epoch 1/3
60000/60000 [==============================] - 335s 6ms/step - loss: 14.4301 - acc: 0.1046 - val_loss: 14.4612 - val_acc: 0.1028
Epoch 2/3
60000/60000 [==============================] - 347s 6ms/step - loss: 14.4351 - acc: 0.1044 - val_loss: 14.4612 - val_acc: 0.1028
Epoch 3/3
60000/60000 [==============================] - 338s 6ms/step - loss: 14.4351 - acc: 0.1044 - val_loss: 14.4612 - val_acc: 0.1028


Above training had ~10.4% accuracy and loss function of 14.4 for the entirety of epoch 1.

~10.4% accuracy and loss function of 14.4 for the entirety of epoch 2 too

~10.4% accuracy and loss function of 14.4 for the entirety of epoch 3 too. Hmmm

In [16]:
# Now predict using what looks like junk from above (10.4% accuracy)

print(model.predict(X_test[:4]))
print(' ')
print(y_test[:4])

# Hmmmm Suspect the issue may lie in the depreciated parts of tensorflow 1. Should redo this with TF2 library

[[0. 0. 0. 0. 0. 0. 0. 1. 0. 0.]
 [0. 0. 0. 0. 0. 0. 0. 1. 0. 0.]
 [0. 0. 0. 0. 0. 0. 0. 1. 0. 0.]
 [0. 0. 0. 0. 0. 0. 0. 1. 0. 0.]]
 
[[0. 0. 0. 0. 0. 0. 0. 1. 0. 0.]
 [0. 0. 1. 0. 0. 0. 0. 0. 0. 0.]
 [0. 1. 0. 0. 0. 0. 0. 0. 0. 0.]
 [1. 0. 0. 0. 0. 0. 0. 0. 0. 0.]]
